### PR TITLE
feat(googleads): Google Ads source plugin

### PR DESCRIPTION
**Pull request by candiduslynx**
Implements #8642

### CODE REVIEW

File: plugins/source/googleads/main.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package main
+
+import (
+	"github.com/cloudquery/cloudquery/plugins/source/googleads/resources/plugin"
+	"github.com/cloudquery/plugin-sdk/serve"
+)
+
+const sentryDSN = "" // TODO
```
Comment (disq) [inline plugins/source/googleads/main.go:8]: TODO found

**Review by disq — CHANGES_REQUESTED** [review]
Few things around spec/init

File: plugins/source/googleads/client/hierarchy.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+				res[k] = append(res[k], curr)
+			}
+		}
+	}
+	return res
+}
+
+// dedup makes sure that each value is saved only once
+func dedup(customers map[string][]string) map[string][]string {
+	saw := make(map[string]struct{})
```
[the remark this answers is not in the document]
Comment (candiduslynx) [reply plugins/source/googleads/client/hierarchy.go:123]: we can prealloc, but the real len will be based on the values, not the key amounts

File: plugins/source/googleads/main.go (hunk 1 above)
Comment (candiduslynx) [reply plugins/source/googleads/main.go:8]: I'll do this in a bit

File: plugins/source/googleads/client/hierarchy.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+	ids := make([]string, len(resp.ResourceNames))
+	for i, rn := range resp.ResourceNames {
+		ids[i] = strings.TrimPrefix(rn, "customers/")
+	}
+
+	return ids, nil
+}
+
+func limitTo(customers map[string][]string, only []string) map[string][]string {
```
[the remark this answers is not in the document]
Comment (candiduslynx) [reply plugins/source/googleads/client/hierarchy.go:88]: What we actually do here is not a dedup per se.

We have `customers` map representing the whole management tree accessible via credentials supplied (possibly limited by `Spec.LoginCustomerID`, but that's another story). This function traverses this tree to add all subtrees that have root matching `only` parameter (if it's empty we simply fetch everything).

The dedup func that is used afterwards is just making sure that we'll fetch the customer data only once (if we want to receive the data from 2 management accounts `A` and `B` and both of them have access to the customer `C` we'll add either `A:C` or `B:C` key-value pair, but not both).

File: plugins/source/googleads/client/context.go [hunk 4]
```diff
@@ -13,21 +13,13 @@ const (
 
 func (c *Client) OutgoingContext(ctx context.Context) context.Context {
 	ctx = c.addDeveloperToken(ctx)
-	return c.addLoginCustomerID(ctx)
+	return addLoginCustomerID(ctx, c.ManagerID)
```
Comment (disq) [inline plugins/source/googleads/client/context.go:16]: Forgot to add `if len(c.ManagerID) > 0 { ...`

Comment (candiduslynx) [reply plugins/source/googleads/client/context.go:16]: I took a look at the logic and the manager ID is always present, so omitted intentionally.

Comment (candiduslynx) [reply plugins/source/googleads/client/context.go:16]: But I'll add it to be consistent

File: plugins/source/googleads/client/context.go [hunk 5]
```diff
@@ -13,6 +13,9 @@ const (
 
 func (c *Client) OutgoingContext(ctx context.Context) context.Context {
 	ctx = c.addDeveloperToken(ctx)
+	if len(c.ManagerID) == 0 {
```
Comment (disq) [inline plugins/source/googleads/client/context.go:16]: If it's consistently there then why add it?

**Review by disq — APPROVED** [review]

File: plugins/source/googleads/client/context.go (hunk 5 above)
Comment (candiduslynx) [reply plugins/source/googleads/client/context.go:16]: Well, it was me who implemented this initially.
If someone then updates the code without this knowledge, they might end up setting the `login-customer-id` to `""`, which, in its turn, may produce unexpected API response.
Better err on the safe side.

**Pull request merged**